src/MediatR/Mediator.cs

```diff
@@ -41,18 +41,28 @@ public Task<TResponse> Send<TResponse>(IRequest<TResponse> request, Cancellation
         }
 
         public Task Publish<TNotification>(TNotification notification, CancellationToken cancellationToken = default)
-            where TNotification : INotification
+             where TNotification : INotification
         {
             if (notification == null)
             {
                 throw new ArgumentNullException(nameof(notification));
             }
 
-            var notificationType = notification.GetType();
-            var handler = _notificationHandlers.GetOrAdd(notificationType,
-                t => (NotificationHandlerWrapper)Activator.CreateInstance(typeof(NotificationHandlerWrapperImpl<>).MakeGenericType(notificationType)));
+            return Publish(notification, typeof(TNotification));

```

Comment:
My bad. Sorry last request - as @remcoros pointed out, using `typeof(TNotification)` would be a regression. Sorrryyyyyy

Reply:
So we'd do `notification.GetType()` here instead. Sorry again!

Reply:
I'll include some tests for this scenario :)